pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotImplicitTableHintRule.java

```diff
@@ -0,0 +1,316 @@
+      Map<String, String> kvOptions) {
+    String partitionFunction = kvOptions.get(kvOptions.get(PinotHintOptions.TableHintOptions.PARTITION_FUNCTION));
+    if (partitionFunction != null && !partitionFunction.equals(tablePartitionInfo.getPartitionFunctionName())) {
+      LOGGER.debug("Override implicit table hint for {} with explicit partition function: {}",
+          tableScan, partitionFunction);
+      return partitionFunction;
+    } else {
+      return tablePartitionInfo.getPartitionFunctionName();
+    }
+  }
+
+  /**
+   * Get the partition parallelism from the hint, if any, otherwise returns null.
+   */
+  @Nullable
+  private static Integer getPartitionParallelism(LogicalTableScan tableScan, Map<String, String> kvOptions) {
+    String partitionParallelismStr = kvOptions.get(PinotHintOptions.TableHintOptions.PARTITION_PARALLELISM);
+    if (partitionParallelismStr != null) {
+      try {
+        int partitionParallelism = Integer.parseInt(partitionParallelismStr);
+        LOGGER.debug("Override implicit table hint for {} with explicit partition parallelism: {}", tableScan,
+            partitionParallelism);
+        return partitionParallelism;
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException(
+            "Invalid partition parallelism: " + partitionParallelismStr + " for table: " + tableScan);
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Get the partition size from the hint, if any, otherwise use the partition size from the table partition info.
+   */
+  private static int getPartitionSize(TablePartitionInfo tablePartitionInfo, LogicalTableScan tableScan,
+      Map<String, String> kvOptions) {
+    int partitionSize = tablePartitionInfo.getNumPartitions();
+    String partitionSizeStr = kvOptions.get(PinotHintOptions.TableHintOptions.PARTITION_SIZE);
+    if (partitionSizeStr != null) {
+      try {
+        int explicitPartitionSize = Integer.parseInt(partitionSizeStr);
+        if (explicitPartitionSize != partitionSize) {
+          LOGGER.debug("Override implicit table hint for {} with explicit partition size: {}", tableScan,
+              explicitPartitionSize);
+          partitionSize = explicitPartitionSize;
+        }
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException("Invalid partition size: " + partitionSizeStr + " for table: " + tableScan);
+      }
+    }
+    return partitionSize;
+  }
+
+  /**
+   * Get the table partition info for the given table scan, transforming Calcite's table name to Pinot's table name
+   * and adding the table type if needed.
+   */
+  @Nullable
+  private TablePartitionInfo getTablePartitionInfo(LogicalTableScan tableScan) {
+    String tableName = RelToPlanNodeConverter.getTableNameFromTableScan(tableScan);
+
+    String offlineName = TableNameBuilder.forType(TableType.OFFLINE).tableNameWithType(tableName);
+    TablePartitionInfo offlineTpi = _workerManager.getTablePartitionInfo(offlineName);
```

Comment:
One possible scenario is that both tables exist, but only one table has partition info. In that case we should return `null` instead

Reply:
Shouldn't we just simplify this with `WorkerManager.calculatePartitionTableInfo(tableName)` now that we have that function?

Reply:
I've changed this code to delegate on WorkerManager. The rule is now simpler and all logic (but merge) is in WorkerManager now.